src/pybind/mgr/rook/rook_cluster.py

```diff
@@ -16,32 +17,163 @@
+        t.start()
+        return t
+
+    return wrapper
+
+# TODO remove +++++++++++++++++++++++++++++++++++
+
+from typing import TypeVar, Generic, ValuesView, Dict
+
+T = TypeVar('T')
+from kubernetes.client import V1ConfigMap, V1ObjectMeta, V1Pod, V1PodSpec, V1Node
+
+
+class KubernetesResource(Generic[T]):
+    def __init__(self, api_func, **kwargs):
+        """
+        Generic kubernetes Resource parent class
+
+        The api fetch and watch methods should be common across resource types,
+
+        Exceptions in the runner thread are propagated to the caller.
+
+        :param api_func: kubernetes client api function that is passed to the watcher
+        :param filter_func: signature: ``(Item) -> bool``.
+        """
+        self.kwargs = kwargs
+        self.api_func = api_func
+        self._items = dict()  # type: Dict[str, T]
+        self.thread = None  # type: threading.Thread
+        self.exception = None
+
+    def _fetch(self):
+        """ Execute the requested api method as a one-off fetch"""
+        response = self.api_func(**self.kwargs)
+        # metadata is a V1ListMeta object type
+        metadata = response.metadata  # type: V1ListMeta
+        self._items = {item.metadata.name: item for item in response.items}
+        return metadata.resource_version
+
+    @property
+    def items(self) -> ValuesView[T]:
+        """
+        Returns the items of the request.
+        Creates the watcher as a side effect.
+        :return:
+        """
+        if self.exception:
+            e = self.exception
+            self.exception = None
+            raise e  # Propagate the exception to the user.
+        if not self.thread or self.thread.is_alive():
+            # Start a thread which will use the kubernetes watch client against a resource
+            resource_version = self._fetch()
+            log.debug("Attaching resource watcher for k8s {}".format(self.api_func))
+            self.thread = self._watch(resource_version)  # type: threading.Thread
+
+        return self._items.values()
+
+    @threaded
+    def _watch(self, res_ver):
+        """ worker thread that runs the kubernetes watch """
+
+        self.exception = None
+
+        w = watch.Watch()
+
+        try:
+            # execute generator to continually watch resource for changes
+            for event in w.stream(self.api_func, resource_version=res_ver, watch=True,
+                                  **self.kwargs):
+                self.health = ''
+                item = event['object']
+                try:
+                    name = item.metadata.name
+                except AttributeError:
+                    raise AttributeError(
+                        "{} doesn't contain a metadata.name. Unable to track changes".format(
+                            self.api_func))
+
+                log.warning('Update event: ' + str(event))
+
+                if event['type'] == 'ADDED':
+                    self._items[name] = item
```

Comment:
Should we add a lock for `_items`? Although it's generally safe to manipulate dict between python threads because of GIL.

Reply:
fixed. now I assume that assignments are atomic. I'm no longer modifying `_items`